FFmpegInterop/NALUSampleProvider.cpp

```diff
@@ -0,0 +1,255 @@
+			{
+				// Reserve additional space since NALU start code is longer than NALU length size
+				sampleSize += MAX_NALU_NUM_SUPPORTED * (sizeof(NALU_START_CODE) - m_naluLengthSize);
+			}
+
+			buf.reserve(sampleSize);
+		}
+
+		vector<uint32_t> naluLengths;
+		naluLengths.reserve(MAX_NALU_NUM_SUPPORTED);
+
+		const uint8_t naluPrefixLength{ m_isBitstreamAnnexB ? sizeof(NALU_START_CODE) : m_naluLengthSize };
+		for (uint32_t i{ 0 }; i + naluPrefixLength <= static_cast<uint32_t>(packet->size);)
+		{
+			// Get the length of the next NALU
+			uint32_t naluLength{ 0 };
+			if (m_isBitstreamAnnexB)
+			{
+				naluLength = GetAnnexBNaluLength(packet->data + i, packet->size - i);
+			}
+			else
+			{
+				naluLength = GetAVCNaluLength(packet->data + i, packet->size - i, m_naluLengthSize);
+			}
+
+			if (copySpsPpsData)
+			{
+				// Check if this NALU is an AUD (access unit delimiter). The AUD must be the first NALU in the sample.
+				if (naluLength < naluPrefixLength || packet->data[naluPrefixLength] != NALU_TYPE_AUD)
+				{
+					// Copy the SPS/PPS data
+					buf.insert(buf.end(), m_spsPpsData.data(), m_spsPpsData.data() + m_spsPpsData.size());
+
+					// Save the SPS/PPS NALU lengths
+					naluLengths.insert(naluLengths.end(), m_spsPpsNaluLengths.begin(), m_spsPpsNaluLengths.end());
+
+					copySpsPpsData = false;
+				}
+			}
+
+			if (writeToBuf)
+			{
+				// Write the NALU start code and data
+				const uint8_t* naluData{ packet->data + i + naluPrefixLength };
+				buf.insert(buf.end(), begin(NALU_START_CODE), end(NALU_START_CODE));
+				buf.insert(buf.end(), naluData, naluData + naluLength);
+			}
+			else if (!m_isBitstreamAnnexB)
+			{
+				// Replace the NALU length with the NALU start code
+				copy(begin(NALU_START_CODE), end(NALU_START_CODE), packet->data + i);
+			}
+
+			// Save the NALU length
+			naluLengths.push_back(sizeof(NALU_START_CODE) + naluLength);
+
+			i += naluPrefixLength + naluLength;
+		}
+
+		if (copySpsPpsData)
+		{
+			// Copy the SPS/PPS data
+			buf.insert(buf.end(), m_spsPpsData.data(), m_spsPpsData.data() + m_spsPpsData.size());
+
+			// Save the SPS/PPS NALU lengths
+			naluLengths.insert(naluLengths.end(), m_spsPpsNaluLengths.begin(), m_spsPpsNaluLengths.end());
+
+			copySpsPpsData = false;
+		}
+
+		if (writeToBuf)
+		{
+			return { make<FFmpegInteropBuffer>(move(buf)), move(naluLengths) };
+		}
+		else
+		{
+			return { make<FFmpegInteropBuffer>(move(packet)), move(naluLengths) };
+		}
+	}
+
+	AnnexBParser::AnnexBParser(_In_reads_(dataSize) const uint8_t* data, _In_ uint32_t dataSize) :
+		m_data(data),
+		m_dataSize(dataSize)
+	{
+		// Validate parameters
+		WINRT_ASSERT(m_data != nullptr);
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, m_dataSize < sizeof(NALU_START_CODE));
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, !equal(begin(NALU_START_CODE), end(NALU_START_CODE), m_data));
+	}
+
+	tuple<vector<uint8_t>, vector<uint32_t>> AnnexBParser::GetSpsPpsData() const
```

Comment:
>GetSpsPpsData [](start = 56, length = 13)

nit: This method implementation isn't SPS / PPS data specific, so perhaps it would be better to use a more generic name here (e.g GetNaluData()). This applies to the spsPpsData / spsPpsNaluLengths local variables in this method as well.